src/control/server/harness.go

```diff
@@ -256,20 +254,96 @@ func (h *IOServerHarness) Start(parent context.Context, membership *system.Membe
 		}
 	}
 
-	// now monitor them
-	select {
-	case <-parent.Done():
-		return parent.Err()
-	case err := <-errChan:
-		// If we receive an error from any instance, shut them all down.
-		// TODO: Restart failed instances rather than shutting everything
-		// down.
-		h.log.Errorf("instance error: %s", err)
-		if err != nil {
-			return errors.Wrap(err, "Instance error")
+	return nil
+}
+
+// monitorInstances listens for exit results from instances or harness and will
+// return only when all harness instances are stopped and restartInstances
+// signal is received.
+func (h *IOServerHarness) monitorInstances(ctx context.Context) error {
+	h.log.Debug("monitoring instances")
+	for {
+		select {
+		case <-ctx.Done(): // received when harness is exiting
+			h.log.Debug("harness exiting")
+			return ctx.Err()
+		case err := <-h.errChan: // received when instance exits
+			// TODO: Restart failed instances on unexpected exit.
+			allInstancesStopped := !h.HasStartedInstances()
+			msg := fmt.Sprintf("instance exited: %v", err)
+			if allInstancesStopped {
+				msg += ", all instances stopped!"
+			}
+			h.log.Info(msg)
+		case <-h.restartInstances: // trigger harness to restart instances
+			h.log.Debug("restart instances")
+			if h.HasStartedInstances() {
+				return errors.New("cannot restart when instances are running")
+			}
+			return nil
+		}
+	}
+
+	return nil
+}
+
+// Start starts all configured instances, waits for them to be ready and then
+// loops monitoring instance exit, harness exit and harness restart signals.
+func (h *IOServerHarness) Start(parent context.Context, membership *system.Membership) error {
+	if h.IsStarted() {
+		return errors.New("can't start: harness already started")
+	}
+
+	// Now we want to block any RPCs that might try to mess with storage
+	// (format, firmware update, etc) before attempting to start I/O servers
+	// which are using the storage.
+	h.Lock()
+	h.started = true
+	h.Unlock()
+
+	ctx, shutdown := context.WithCancel(parent)
+	defer shutdown()
+
+	for {
+		if err := h.startInstances(ctx, membership); err != nil {
+			return err
+		}
+		if err := h.waitInstancesReady(ctx); err != nil {
+			return err
+		}
+		if err := h.monitorInstances(ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// HasStartedInstances returns true if any harness instances are running.
+func (h *IOServerHarness) HasStartedInstances() bool {
+	h.RLock()
+	defer h.RUnlock()
+
+	for _, instance := range h.instances {
+		if instance.IsStarted() {
+			return true
 		}
 	}
 
+	return false
+}
+
+// RestartInstances will signal the harness to restart configured instances.
+func (h *IOServerHarness) RestartInstances() error {
+	if !h.IsStarted() {
+		return errors.New("can't start instances: harness not started")
+	}
+	if h.HasStartedInstances() {
+		return errors.New("can't start instances: already started")
+	}
+
+	h.restartInstances <- struct{}{} // trigger harness to restart its instances
```

Comment:
How will this method be called? If it's in the same goroutine as the one running monitorInstances(), this could result in a deadlock. harness.Start() is being called in a new goroutine from server.go so this is probably OK, but just something to be careful with.

Reply:
moved to https://github.com/daos-stack/daos/pull/1541, will never be the same goroutine as running monitor() because it's only ever called over gRPC in a handler, not by the goroutine spawned from server main thread itself (running Start()).